twister: add option cleanup all tests

congnguyenhuu opened the pull request:
Add 2 choice  for `--runtime-artifact-cleanup` option, delete artifacts of passing tests by choicing 'pass' 
or delete artifacts of both passing tests and failing tests by choicing 'all'.
Saving resources for machine, incase run twister there are many failed test.
Keep testcase_extra.config, it contains some extra Kconfig needed for the tests and they is only generated one time.

congnguyenhuu committed:
twister: add option cleanup all tests

add 2 choices for '--runtime-artifact-cleanup' option,
'passing' to delete artifacts of passing tests,
'all' to delete artifacts of both passing tests
and failing tests

keep testcase_extra.config, it contains some extra Kconfig
needed for the tests and they is only generated first time
when run retry failed

Signed-off-by: Cong Nguyen Huu <cong.nguyenhuu@nxp.com>

hakehuang commented:
@congnguyenhuu what is your user case for this PR? If you run twister, and then remove the result right after the test, why you run it?

congnguyenhuu commented:
> @congnguyenhuu what is your user case for this PR? If you run twister, and then remove the result right after the test, why you run it?

This option is same `--runtime-artifact-cleanup` option help to save resources PC. But it different that cleanup even if test failed.
And it still keep the twister result (keep files: log, config, cmakecache, runner, hex, elf), only delete artifacts of tests.

hakehuang commented:
> This option same as `--runtime-artifact-cleanup` option help to save resources PC. But it different that cleanup even if test failed.
> And it still keep the twister result (keep files: log, config, cmakecache, runner, hex, elf), only delete artifacts of tests.

if in this case, better to modify the --runtime-artifact-cleanup to --runtime-artifact-cleanup=1 keep failure data, --runtime-artifact-cleanup=2 do not keep failure data, would this be better?

congnguyenhuu commented:
> 



> > This option same as `--runtime-artifact-cleanup` option help to save resources PC. But it different that cleanup even if test failed.
> > And it still keep the twister result (keep files: log, config, cmakecache, runner, hex, elf), only delete artifacts of tests.
> 
> if in this case, better to modify the --runtime-artifact-cleanup to --runtime-artifact-cleanup=1 keep failure data, --runtime-artifact-cleanup=2 do not keep failure data, would this be better?

I agree that is a good idea. I will update.

congnguyenhuu commented:
> > This option same as `--runtime-artifact-cleanup` option help to save resources PC. But it different that cleanup even if test failed.
> > And it still keep the twister result (keep files: log, config, cmakecache, runner, hex, elf), only delete artifacts of tests.
> 
> if in this case, better to modify the --runtime-artifact-cleanup to --runtime-artifact-cleanup=1 keep failure data, --runtime-artifact-cleanup=2 do not keep failure data, would this be better?

I updated  with 2 choice for `--runtime-artifact-cleanup` : 'passing-tests' and 'all-tests'. It is more reminiscent than 1,2.

stephanosio commented:
There was a problem with the CI. A fix has been merged, so please rebase.

congnguyenhuu commented:
> There was a problem with the CI. A fix has been merged, so please rebase.

I rebased, a failing CI is fixed. But I still have other failing CI that relate to new update for `--runtime-artifact-cleanup` option.
I updated action twister for `--runtime-artifact-cleanup` with `-M passing-tests` choice in workflow CI. It seems CI is not working with  new option update, it still works with old option `-M`. I don't know clear how CI works, could you for me a suggestion about this? 
![image](https://user-images.githubusercontent.com/76719459/182990871-2e2ba1e4-6829-42fb-b482-6591f0b56897.png)

hakehuang commented:
> > There was a problem with the CI. A fix has been merged, so please rebase.
> 
> I rebased, a failing CI is fixed. But I still have other failing CI that relate to new update for `--runtime-artifact-cleanup` option. I updated action twister for `--runtime-artifact-cleanup` with `-M passing-tests` choice in workflow CI. It seems CI is not working with new option update, it still works with old option `-M`. I don't know clear how CI works, could you please give me a suggestion about this? ![image](https://user-images.githubusercontent.com/76719459/182990871-2e2ba1e4-6829-42fb-b482-6591f0b56897.png)

you need update the github actions scripts as well

stephanosio reviewed (dismissed):
> I updated action twister for --runtime-artifact-cleanup with -M passing-tests choice in workflow CI.

Please make `-M` default to `passing-tests`, instead of requiring everyone to type `passing-tests` every time.

congnguyenhuu commented:
> > I updated action twister for --runtime-artifact-cleanup with -M passing-tests choice in workflow CI.
> 
> Please make `-M` default to `passing-tests`, instead of requiring everyone to type `passing-tests` every time.

I updated. That  also helps me slove the failing CI.

stephanosio commented on .github/workflows/clang.yaml:
```diff
@@ -110,7 +110,7 @@ jobs:
             echo "::set-output name=report_needed::1";
             # Full twister but with options based on changes
-            ./scripts/twister --force-color --inline-logs -M -N -v --load-tests testplan.json --retry-failed 2
+            ./scripts/twister --force-color --inline-logs -M all-tests -N -v --load-tests testplan.json --retry-failed 2
```
why do we need this change?

```suggestion
            ./scripts/twister --force-color --inline-logs -M -N -v --load-tests testplan.json --retry-failed 2
```

congnguyenhuu replied:
> why do we need this change?

Just i want to CI check twister with this option after updating

congnguyenhuu replied:
@stephanosio you mean that CI should check twister with default option` -M`

nashif replied:
yes, please leave that alone please, we will consider enabling this in CI later.

nashif commented on scripts/pylib/twister/twisterlib/environment.py:
```diff
@@ -363,8 +364,11 @@ def parse_arguments(args):
     parser.add_argument(
-        "-M", "--runtime-artifact-cleanup", action="store_true",
-        help="Delete artifacts of passing tests.")
+        "-M", "--runtime-artifact-cleanup", choices=['passing-tests', 'all-tests'],
```
prefer to keep things short, i.e. `choices=["pass", "all"]`

nashif commented on scripts/pylib/twister/twisterlib/environment.py:
```diff
@@ -363,8 +364,11 @@ def parse_arguments(args):
-        "-M", "--runtime-artifact-cleanup", action="store_true",
-        help="Delete artifacts of passing tests.")
+        "-M", "--runtime-artifact-cleanup", choices=['passing-tests', 'all-tests'],
+        default=None, const='passing-tests', nargs='?',
```
Dont we need a default which is compatible with current behavior, ie.. default should be 'pass', how is the default being set? using const?

nashif commented on scripts/pylib/twister/twisterlib/environment.py:
```diff
@@ -363,8 +364,11 @@ def parse_arguments(args):
-        help="Delete artifacts of passing tests.")
+        "-M", "--runtime-artifact-cleanup", choices=['passing-tests', 'all-tests'],
+        default=None, const='passing-tests', nargs='?',
+        help="""This option allows to delete artifacts of tests. The default value
```
`Cleanup test artifacts; The default behavior is 'pass' which only removed artifacts of passing tests, if you wish to remove all artificats including those of failed tests, use 'all'`

nashif commented on scripts/pylib/twister/twisterlib/runner.py:
```diff
@@ -522,11 +524,14 @@ def cleanup_artifacts(self, additional_keep=[]):
         allow += additional_keep
 
+        if self.options.runtime_artifact_cleanup == 'all-tests':
+            allow += [os.path.join('twister', 'testcase_extra.conf')]
```
why is this file special?

nashif commented on scripts/pylib/twister/twisterlib/runner.py:
```diff
@@ -498,22 +499,23 @@ def process(self, pipeline, done, message, lock, results):
                 self.report_out(results)
 
-            if self.options.runtime_artifact_cleanup and not self.options.coverage and self.instance.status == "passed":
+            if (self.options.runtime_artifact_cleanup == 'passing-tests' and not self.options.coverage and self.instance.status == "passed")\
```
this conditional is becoming unreadable and too long, please reformat or simplify/split the logic.

nashif commented on scripts/pylib/twister/twisterlib/environment.py:
```diff
@@ -365,8 +366,11 @@ def parse_arguments(args):
+        "-M", "--runtime-artifact-cleanup", choices=['pass', 'all'],
+        default=None, const='pass', nargs='?',
+        help="""Cleanup test artifacts. The default behavior is 'pass'
+        which only removed artifacts of passing tests. If you wish to
```
s/removed/removes

congnguyenhuu replied:
I updated to "removes"

congnguyenhuu commented:
@stephanosio, @hakehuang  pls help me review this, all comments are addressed
